### PR TITLE
Feature/tracker

### PR DESCRIPTION
**Pull request by EmiM**

### CODE REVIEW

File: src/tracker/index.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  private setRouting() {
+    this._app.use(express.json())
+    this._app.get('/peers', (req, res) => {
+      res.send(this.getPeers())
+    })
+    this._app.post('/register',(req, res) => {
+      console.log('body', req.body)
+      const address = req.body['address']
```
Comment (kowalski) [inline src/tracker/index.ts:66]: we should also store expirationTime of each entry we add.

File: src/tracker/index.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    await tor.init()
+    return await tor.addOnion({ virtPort: this._port, targetPort: this._port, privKey: process.env.HIDDEN_SERVICE_SECRET })
+  }
+
+  private addPeer(address: string) {
+    try {
+      multiaddr(address)
+    } catch (e) {
+      console.debug('Wrong address format:', e)
+      return
```
Comment (kowalski) [inline src/tracker/index.ts:49]: this should result in 400 response

File: src/tracker/index.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      res.end()
+    })
+  }
+
+  public async init() {
+    await this.initTor()
+    this.setRouting()
+  }
+
+  public listen() {
```
Comment (kowalski) [inline src/tracker/index.ts:81]:
```
async public listen(): Promise<void> {
  return new Promise(resolve => {
    this._app.listen(this._port, () => {
      console.debug(`Tracker listening on ${this._port}`)
      resolve()
    })
    }
    }
    ```
    
    This way it's easier to control - you get a Promise that tells you when server is listening and ready

File: src/tracker/index.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      options: {
+        env: {
+          LD_LIBRARY_PATH: pathDevLib,
+          HOME: os.homedir()
+        },
+        detached: true
+      }
+    })
+    await tor.init()
+    return await tor.addOnion({ virtPort: this._port, targetPort: this._port, privKey: process.env.HIDDEN_SERVICE_SECRET })
```
Comment (kowalski) [inline src/tracker/index.ts:41]: preferably you would also get `privKey` i constructor. This way this class is not making assumptions about how secret is passed in

File: src/tracker/index.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  private clearPeers() {
+    const now = (new Date()).getTime()
+    for (const address of Object.keys(this._peers)) {
+      if (now > this._peers[address]) {
+        delete this._peers[address]
+      }
+    }
+  }
```
Comment (karim-agha) [inline src/tracker/index.ts:77]: I'd argue that a cleaner and more performant approach would be to store the set of peers in [LruSet](https://www.collectionsjs.com/lru-set) or [SortedSet](https://www.collectionsjs.com/sorted-set) and prune to bottom of the list periodically.

**Pull request merged**